src/optimizer/topn_optimizer.cpp

```diff
@@ -20,13 +19,41 @@ bool TopN::CanOptimize(LogicalOperator &op) {
 			// we need offset to be either not set (i.e. limit without offset) OR have offset be
 			return false;
 		}
-		return true;
+
+		auto child_op = op.children[0].get();
+
+		while (child_op->type == LogicalOperatorType::LOGICAL_PROJECTION) {
+			D_ASSERT(!child_op->children.empty());
+			child_op = child_op->children[0].get();
+		}
+
+		return child_op->type == LogicalOperatorType::LOGICAL_ORDER_BY;
 	}
 	return false;
 }
 
 unique_ptr<LogicalOperator> TopN::Optimize(unique_ptr<LogicalOperator> op) {
 	if (CanOptimize(*op)) {
+
+		vector<unique_ptr<LogicalOperator>> projections;
+
+		if (op->children[0]->type != LogicalOperatorType::LOGICAL_ORDER_BY) {
```

Comment:
something like might look nicer. If there is no projection sitting in the way, the while loop won't run. You already know you can optimize, so the cast will be valid (and the `D_ASSERT` will check it in debug mode.
```
auto child = std::move(op->children[0]);
// collect all projections until we get to the order by
while (child->type == LogicalOperatorType::LOGICAL_PROJECTION) {
	D_ASSERT(!child->children.empty());
	auto tmp = std::move(child->children[0]);
	projections.push_back(std::move(child));
	child = std::move(tmp);
}
D_ASSERT(child->type == LogicalOperatorType::LOGICAL_ORDER_BY);
auto &order_by = child->Cast<LogicalOrder>();
```
might look nicer.

Reply:
Awesome, that looks good to me.